### DISCUSSION TITLE
Is it possible to override the max_clients number in simulator mode?

### DISCUSSION THREAD

**Discussion by magdalena-fuchs** [NVIDIA/NVFlare#1454] [Q&A] [answered]
Ideally, we would like to simulate 500+ clients but, by default, the maximum number of clients for the simulator runner is 100. 
Is it possible to define a custom simulator runner or to override this "max_clients" default in any other way? 

And is there a good reason not to (other than time and resources)?

**Comment by YuanTingHsieh** [comment, marked as answer]
We just picked a large enough number as default there.
When you are writing codes, you can change that value when create you `SimulatorRunner`.
```
simulator = SimulatorRunner(
        job_folder=simulator_args.job_folder,
        workspace=simulator_args.workspace,
        clients=simulator_args.clients,
        n_clients=simulator_args.n_clients,
        threads=simulator_args.threads,
        gpu=simulator_args.gpu,
        max_clients=500,
    )
```

When you are using command line you can specify:
```
nvflare simulator -m 500
```

If you want to have a custom class that overrides the default one way is:
```
class CustomSimulatorRunner(SimulatorRunner):
    def __init__(self, other args here):
        super(args1 = args1, args2=args2, max_clients=500)
```

As far as I know, you can change that number for your system, only limit is resource.

**Reply by magdalena-fuchs** [reply]
Thank you, that was really helpful! I hadn't realized I could simply change the max from the CLI command.

**Reply by magdalena-fuchs** [reply]
Actually, going off of this topic: Can I also override classes like the Communicator (similar to what you show with `CustomSimulatorRunner`)? If so, how would I configure this in the config json for my clients to use the custom class instead of the provided one?

**Reply by YuanTingHsieh** [reply]
@magdalena-fuchs 

No problem at all, you could theoretically overrides/replaces any classes in our system you see fit.

We can't customize that using config json for now, you can overwrite those classes and build a custom wheel for your clients.
